### PR TITLE
[Draft] Add support for Json Google Cloud Storage Data Connector

### PR DESCRIPTION
**Pull request by i25959341**
## Notice

- [x] I acknowledge and agree that, by checking this box and clicking "Submit Pull Request":

- I submit this contribution under the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt) and represent that I am entitled to do so on behalf of myself, my employer, or relevant third parties, as applicable.
- I certify that (a) this contribution is my original creation and / or (b) to the extent it is not my original creation, I am authorised to submit this contribution on behalf of the original creator(s) or their licensees.
- I certify that the use of this contribution as authorised by the Apache 2.0 license does not violate the intellectual property rights of anyone else.

## Motivation and Context
Terrible PR, but was throwing rocks to bait the jades.

This was created as a draft PR for #108, #109, #110, #111 . It adds support for Google Cloud storage Parquet/Json/pickle/hdf/csv datasets. 
This work is heavily inspired by @plauto and his work at #100 

## How has this been tested?
I have used the same testing strategies as @plauto as his work at #100. I would like some feedbacks on how to parametrise the test based on the file format thus using just one test file instead many.

Also not a fan of these YAML files. Maybe worth moving back to google cloud python client instead? Not sure on that

Really a fun PR, great to learn something about VCR @plauto

## Checklist

- [ ] Read the [contributing](/CONTRIBUTING.md) guidelines
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added a description of this change and added my name to the list of supporting contributions in the [`RELEASE.md`](/RELEASE.md) file
- [ ] Added tests to cover my changes
- [ ] Assigned myself to the PR

### CODE REVIEW

**Comment by i25959341** [comment]
@921kiyo Hey kiyo, added versioning tests

File: tests/contrib/io/gcs/test_json_gcs.py [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+    def test_invalid_credentials(self):
+        """Test invalid credentials for connecting to GCS"""
+        pattern = "Anonymous caller"
+        with pytest.raises(DataSetError, match=pattern):
+            JsonGCSDataSet(filepath=FILENAME, bucket_name=BUCKET_NAME).load()
+
+    @gcs_vcr.use_cassette(match=["api_recordings/json/*.yaml"])
+    def test_not_existing_bucket(self):
+        """Test not existing bucket"""
+        pattern = "Failed while loading data from data set"
```
Comment (921kiyo) [inline tests/contrib/io/gcs/test_json_gcs.py]:
```suggestion
        pattern = r"Failed while loading data from data set JSONGCSDataSet\(.+\)"
```

File: tests/contrib/io/gcs/test_json_gcs.py [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+        save_args=save_args,
+        version=Version(load_version, save_version),
+    )
+
+
+class TestJsonGCSDataSetVersioned:
+    @gcs_vcr.use_cassette(match=["api_recordings/json/*.yaml"])
+    def test_no_versions(self, versioned_gcs_data_set):
+        """Check the error if no versions are available for load."""
+        pattern = r"Did not find any versions for"
```
Comment (921kiyo) [inline tests/contrib/io/gcs/test_json_gcs.py]:
```suggestion
        pattern = r"Did not find any versions for JSONGCSDataSet\(.+\)"
```

File: tests/contrib/io/gcs/utils.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+# Trademarks or any confusingly similar mark as a trademark for your product,
+#     or use the QuantumBlack Trademarks in any other manner that might cause
+# confusion in the marketplace, including but not limited to in advertising,
+# on websites, or on software.
+#
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+def matcher(r1, r2):  # pylint: disable=too-many-return-statements
```
Comment (921kiyo) [inline tests/contrib/io/gcs/utils.py]: Could you add docstring for this util function, and add a reference where you got this func?

**Comment by i25959341** [comment]
Thanks for the comment, will update

File: tests/contrib/io/gcs/api_recordings/json/test_existed_versioned.yaml [hunk 4]
```diff
@@ -0,0 +1,615 @@
+interactions:
```
[the remark this answers is not in the document]
Comment (i25959341) [reply tests/contrib/io/gcs/api_recordings/json/test_existed_versioned.yaml]: This is generated from real interaction with GCP and recorded by the vcr library

File: requirements.txt [hunk 5]
```diff
@@ -1,6 +1,7 @@
 anyconfig>=0.9.7, <1.0
 click<8.0
 cookiecutter>=1.6.0, <2.0
+gcsfs>=0.3.0, <1.0
```
[the remark this answers is not in the document]
Comment (karlsonthisisbud) [reply requirements.txt]: Updated

File: kedro/contrib/io/gcs/json_gcs.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+        data = data.to_json(**self._save_args)
+
+        with self._gcs.open(str(save_path), mode="wb") as gcs_file:
+            gcs_file.write(data.encode("utf8"))
+
+        # gcs maintain cache of the directory,
+        # so invalidate to see new files
+        self._gcs.invalidate_cache()
+        load_path = PurePosixPath(self._get_load_path())
+        self._check_paths_consistency(load_path, save_path)
```
[the remark this answers is not in the document]
Comment (karlsonthisisbud) [reply kedro/contrib/io/gcs/json_gcs.py]: Updated

File: tests/contrib/io/gcs/test_json_gcs.py (hunk 1 above)
Comment (karlsonthisisbud) [reply tests/contrib/io/gcs/test_json_gcs.py]: Updated

File: tests/contrib/io/gcs/test_json_gcs.py (hunk 2 above)
Comment (karlsonthisisbud) [reply tests/contrib/io/gcs/test_json_gcs.py]: Updated

File: kedro/contrib/io/gcs/json_gcs.py [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+
+    def _exists(self) -> bool:
+        try:
+            load_path = self._get_load_path()
+        except DataSetError:
+            return False
+
+        return self._gcs.exists(load_path)
```
[the remark this answers is not in the document]
Comment (karlsonthisisbud) [reply kedro/contrib/io/gcs/json_gcs.py]: Updated

File: kedro/contrib/io/gcs/json_gcs.py [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+# license to the QuantumBlack Trademarks. You may not use the QuantumBlack
+# Trademarks or any confusingly similar mark as a trademark for your product,
+#     or use the QuantumBlack Trademarks in any other manner that might cause
+# confusion in the marketplace, including but not limited to in advertising,
+# on websites, or on software.
+#
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""``JsonGCSDataSet`` loads and saves data to a file in gcs. It uses google-cloud-storage
```
[the remark this answers is not in the document]
Comment (karlsonthisisbud) [reply kedro/contrib/io/gcs/json_gcs.py]: Updated

File: kedro/contrib/io/gcs/json_gcs.py [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+from typing import Any, Dict, Optional
+
+import gcsfs
+import pandas as pd
+from google.auth.credentials import Credentials
+
+from kedro.io.core import AbstractVersionedDataSet, DataSetError, Version
+
+
+class JsonGCSDataSet(AbstractVersionedDataSet):
```
[the remark this answers is not in the document]
Comment (karlsonthisisbud) [reply kedro/contrib/io/gcs/json_gcs.py]: Updated

File: kedro/contrib/io/gcs/json_gcs.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (i25959341) [reply kedro/contrib/io/gcs/json_gcs.py]: Done

File: kedro/contrib/io/gcs/json_gcs.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (i25959341) [reply kedro/contrib/io/gcs/json_gcs.py]: Done

File: kedro/contrib/io/gcs/json_gcs.py [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+        >>>                          bucket_name="test_bucket",
+        >>>                          load_args=None,
+        >>>                          save_args={"index": False})
+        >>> data_set.save(data)
+        >>> reloaded = data_set.load()
+        >>>
+        >>> assert data.equals(reloaded)
+    """
+
+    DEFAULT_LOAD_ARGS = {}  # type: Dict[str, Any]
```
[the remark this answers is not in the document]
Comment (i25959341) [reply kedro/contrib/io/gcs/json_gcs.py]: Done

File: kedro/contrib/io/gcs/json_gcs.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (i25959341) [reply kedro/contrib/io/gcs/json_gcs.py]: Done

File: requirements.txt (hunk 5 above)
[the remark this answers is not in the document]
Comment (i25959341) [reply requirements.txt]: Done

File: kedro/contrib/io/gcs/json_gcs.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (i25959341) [reply kedro/contrib/io/gcs/json_gcs.py]: Done

File: tests/contrib/io/gcs/api_recordings/json/test_existed_versioned.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (i25959341) [reply tests/contrib/io/gcs/api_recordings/json/test_existed_versioned.yaml]: Done

File: tests/contrib/io/gcs/test_json_gcs.py (hunk 1 above)
Comment (i25959341) [reply tests/contrib/io/gcs/test_json_gcs.py]: Done

File: tests/contrib/io/gcs/test_json_gcs.py (hunk 2 above)
Comment (i25959341) [reply tests/contrib/io/gcs/test_json_gcs.py]: Done

File: tests/contrib/io/gcs/utils.py (hunk 3 above)
Comment (921kiyo) [reply tests/contrib/io/gcs/utils.py]: Plus, as CI says, you need to add unit test for this function to meet 100% coverage.

Comment (i25959341) [reply tests/contrib/io/gcs/utils.py]: yup will get around to it lol

**Comment by yetudada** [comment]
@i25959341 We picked up this PR and thanks to you we'll have a `JSONGCSDataSet` in the next Kedro release. Look out for `kedro 0.15.5` and thank you so much for helping us with this. I'm going to close this PR in anticipation of the release.

**Pull request closed without merging**